No longer passes missing in list compound engine

brejoc opened the pull request:
### What does this PR do?

Missing targets cause a very odd behaviour in combination with a "not" statement and don't seem to be used in any other case.

Having a compound list target like `not L@foominion` would cause a return for `foominion` even though it doesn't exist. Returning an empty list for "missing" fixes that.

### What issues does this PR fix or reference?

https://github.com/saltstack/salt/issues/49205

### Previous Behavior

Given that there is only one minion registered called `minion1`.
```
$ salt -C 'not L@foominion' grains.get os
minion1:
    SUSE
foominion:
    Minion did not return. [No response]
```

### New Behavior
Given that there is only one minion registered called `minion1`.
```
$ salt -C 'not L@foominion' grains.get os
minion1:
    SUSE
```

### Tests written?
No

### Commits signed with GPG?

Yes

brejoc committed:
No longer passes missing in list compound engine

Missing targets cause a very odd behaviour in combination with a
"not" statement and don't seem to be used in every other case.

Having a compound list target like "not L@foominion" would cause
a return for "foominon" even though it doesn't exist. Returning
an empty list for "missing" fixes that.

gtmanfred commented on salt/utils/minions.py:
```diff
@@ -224,7 +224,7 @@ def _check_list_minions(self, expr, greedy):  # pylint: disable=unused-argument
             expr = [m for m in expr.split(',') if m]
         minions = self._pki_minions()
         return {'minions': [x for x in expr if x in minions],
-                'missing': [x for x in expr if x not in minions]}
```
This is not ok.

This was a specific feature added, and would rip out the behavior that shows regular list minions aren't specified.

`L@MINION_NOT_SPECIFIED` would not return that minion_not_specified is not a minion anymore.

brejoc replied:
That is odd. With an unmodified version of Salt (2018.3.0-26.1 from SUSE) I see this:
```
salt -C 'L@foo' grains.get os
No minions matched the target. No command was sent, no jid was assigned.
ERROR: No return received
```
So you'd expect to get a return from the missing minion here or did I misunderstand you? Thanks!

brejoc replied:
See [new comment](https://github.com/saltstack/salt/pull/49435#issuecomment-417671177) please.

brejoc committed:
Ignore missing minions only when excluding them with 'not'

Co-authored-by: Mihai Dinca <Mihai.Dinca@suse.com>

brejoc commented:
@gtmanfred I think we found the use-case you where referring to. You where probably talking about a list like `L@minion1,foominion`, which should also return a response for `foominion` even though it is not existing. We've push an update that should reflect that.

cachedout commented:
@brejoc Right now, @gtmanfred is out for a little while on leave but he should be back in a few weeks to look at this. I'm not totally positive which case he's referring to so I'd rather wait and let him respond instead of guessing. Apologies for the delay.

brejoc commented:
Perhaps @garethgreenaway can take a look here, since he is afaik the original author of the change that introduced this behavior? Sorry for the impatience, but this is blocking the CaaSP release. Thanks!

Commit:
Merge branch 'develop' into develop-fix-for-list-compound-targeting

garethgreenaway committed:
Merge branch 'develop' into develop-fix-for-list-compound-targeting

Commit:
Merge branch 'develop' into develop-fix-for-list-compound-targeting

rallytime commented:
@brejoc This is causing some tests to fail: https://jenkinsci.saltstack.com/job/pr-kitchen-ubuntu1604-py2/job/PR-49435/5/

Can you take a look there? The matching and batch tests are related to this change.

brejoc committed:
Minor fix for def test_batch_run_grains_targeting

item was never "minion" without a colon, but os_grain was overwritten
anyway. So this fix does not actually affect the test outcome.

brejoc committed:
Adds check for engine before adding the additional engine argument

The additional engine argument is only intended to be used with the
_check_list_minons engine and shouldn't be added for every other engine.

brejoc commented:
@rallytime Test should be fixed now. Apart from `integration/shell/test_matcher.py::MatchTest::test_grain`. Looks like this was broken before.

cachedout commented:
@terminalmage Since you have been in the matchers code today, you might have an opinion on this.

terminalmage reviewed (commented):
I would wait until https://github.com/saltstack/salt/pull/48809 is merged, and then re-do this PR. The reason for this is that most of this code will no longer exist (or will exist in a different form, in a different file) once that PR is merged.

brejoc commented:
@terminalmage Sounds good. But are the changes from #48809 backported? Because this fixes a bug that exists down to at least `2018.3.0`.

gtmanfred commented:
This pr and that one is being proposed against develop.

So you will need to open a new pr against the older release to have it be backported.

rallytime commented:
#48809 is against fluorine as making the matchers pluggable is a new feature for fluorine.

terminalmage commented:
@brejoc Where possible, please open pull requests against the oldest release branch where the bug exists: https://docs.saltstack.com/en/latest/topics/development/contributing.html#which-salt-branch

In this case though we would need two PRs due to the massive difference in how matchers are architected beginning in the Fluorine release. You can open one PR against `2018.3` and another against our `fluorine` branch.

cachedout commented:
@gtmanfred Re-review requested.

Commit:
Merge branch 'develop' into develop-fix-for-list-compound-targeting
